### PR TITLE
Pattern-based registrations/subscriptions

### LINKED ISSUE #46: Implement pattern-based subscriptions/registrations

### PR DESCRIPTION
**Pull request by DZabavchik**
Resolves #46, #95, Answers #2 

+ Visual Studio 2015 solution and projects (Updated .gitignore)
+ invocation and event provide actual URI from message to callee/handler.
+ Fixes for test projects (tested with crossbar 0.12.1)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,3 +26,247 @@ doc/_doxygen*
 CMakeLists.txt.user
 *.key
 *.pid
+
```
Comment (oberstet) [inline .gitignore]: the amount of stuff to ignore for VS is ridiculous

File: autobahn/wamp_arguments.hpp [hunk 2]
```diff
@@ -45,6 +45,78 @@ using wamp_kw_arguments = std::unordered_map<std::string, msgpack::object>;
 static const msgpack::object EMPTY_ARGUMENTS(std::array<msgpack::object, 0>(), nullptr);
 static const msgpack::object EMPTY_KW_ARGUMENTS(wamp_kw_arguments(), nullptr);
 
+
+//msgpack map utilities.  
+//TODO: refactor event & invocation to used these
+template <typename T>
+inline T value_for_key(const msgpack::object& object, const std::string& key)
+{
+	if (object.type != msgpack::type::MAP) {
```
Comment (jpetso) [inline autobahn/wamp_arguments.hpp]: Based on the size of indentations that the web interface is showing, it appears that this commit uses tabs to indent rather than spaces. This is inconsistent with the code so far and might lead to weird-looking code in some people's editors.

File: autobahn/wamp_invocation.ipp [hunk 3]
```diff
@@ -193,12 +199,13 @@ inline void wamp_invocation_impl::result(const List& arguments)
     message->set_field(3, arguments);
 
     m_send_result_fn(message);
-    m_send_result_fn = send_result_fn();
+    if (!intermediate)
+		m_send_result_fn = send_result_fn();
```
Comment (jpetso) [inline autobahn/wamp_invocation.ipp]: Code elsewhere in autobahn-cpp is using braces for all `if` statements, including one-liners. That is good practice because it minimizes the potential for future errors when somebody might try to add a second line, and should really be adopted by all C/C++ developers everywhere. At least in this project though.

File: examples/projects/VS2015/autobahn-cpp-examples.vs2015.sln [hunk 4]
```diff
@@ -0,0 +1,84 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (jpetso) [inline examples/projects/VS2015/autobahn-cpp-examples.vs2015.sln]: How should these VS project files be maintained by non-Windows developers, and why is the cross-platform CMake build not sufficient to generate these without committing them?

File: autobahn/wamp_arguments.hpp (hunk 2 above)
Comment (oberstet) [inline autobahn/wamp_arguments.hpp]: I agree. Pls no tabs at all.

File: examples/projects/VS2015/autobahn-cpp-examples.vs2015.sln (hunk 4 above)
Comment (oberstet) [inline examples/projects/VS2015/autobahn-cpp-examples.vs2015.sln]: I agree. The amount of build file insanity for VS is plain ridiculous. cmake should be used, not native VS build crap.

**Comment by oberstet** [comment]
also, to test, I need working examples again first. see https://github.com/crossbario/autobahn-cpp/issues/98

File: autobahn/wamp_invocation.hpp [hunk 5]
```diff
@@ -195,13 +200,13 @@ class wamp_invocation_impl
      * Reply to the invocation with positional arguments.
      */
     template <typename List>
-    void result(const List& arguments);
+    void result(const List& arguments, bool intermediate = false);
```
Comment (jpetso) [inline autobahn/wamp_invocation.hpp]: You're not using the new parameter in any examples in this commit, it's undocumented and I'm not clear what it's purpose is. I'd be hesitant to merge more complicated API without having a clear view of what benefits it brings, and that those benefits outweigh the complexity and loss of readability that an additional boolean brings here. Please explain.

Comment (oberstet) [inline autobahn/wamp_invocation.hpp]: Yep. This PR combines multiple things (and it's easier to merge focused PRs). And it's unclear to be too what `intermediate` is for ..

**Comment by oberstet** [comment]
@DZabavchik all in all, I would like to definitely merge from this PR the following:

* the things for subscribe options
* the guarding for non-existing UDS on Windows

I am also ok with merging all the VS specific stuff like the Git ignore and VS project/build files (because that's probably what Windows dev would expect .. not cmake, though that should work on Windows too).

My main issue hence is with: `bool intermediate`, because it introduces new API.

Could you explain?

Is that supposed to be something for progressive call results?

**Comment by oberstet** [comment]
The merge conflict now is because the subscribe options and CHALLENGE (missing break) bugs were fixed .. only a few conflicting lines.

File: autobahn/wamp_invocation.hpp (hunk 5 above)
Comment (DZabavchik) [inline autobahn/wamp_invocation.hpp]: Will revoke PR, and resubmit as multiple PRs. 

VS project samples can be removed and kept on a fork. But since they are for samples only not the main project no harm in keeping them. 

Sorry about tabs, not a fan of tabs myself. Will fix.

Intermediate is a provision for progressive calls. Will remove and submit as separate PR when complete and tested.

**Comment by DZabavchik** [comment]
I guess it was a case of simultaneous discovery. I went through the all of the samples except for UDS, and made sure they work. Fixed the subscribe, and so did David.

The missing break that was fixed in PR 91, went missing again after epic merge for transport abstraction.

Regards,

Denis

**Comment by oberstet** [comment]
@DZabavchik sounds good! Thanks!

**Comment by DZabavchik** [comment]
@oberstet, 
> re: sounds good!
Should I cancel/close the PR?

**Comment by oberstet** [comment]
@DZabavchik Ah right .. as you said you would file follow up PRs, I closed this one. Pls note that I also just merged @davidchappelle PR https://github.com/crossbario/autobahn-cpp/pull/100 .. if you could pull that so that you don't waste your time with potential conflicts. Thanks for contributing!

**Comment by oberstet** [comment]
@DZabavchik one more note regarding progressive call results. Instead of expanding the invocation callback with individual parameters (like a progressive result callback), we might consider adding just 1 parameter: a call details (and for pubsub, an event details) _object_. That object can then hold a progressive result callback, but can also hold details like caller session ID (for caller disclosure) and the like. More extensible. At least that's how AutobahnPython and AutobahnJS have it ..

**Comment by DZabavchik** [comment]
we still need to differentiate between intermediate results (and set progress:true), and final results. In addition we need to take into consideration whether caller is expecting/requested progressive results.  There is already set_details on invocation that copies URI, same method can look at "receive_progress": true.  Result callback should either discard intermediate results when caller did not request them, or throw.

**Comment by davidchappelle** [comment]
@DZabavchik <https://github.com/DZabavchik> I was also working on adding
support for progressive call results. We probably had something similar I
haven't looked at your branch yet. But if you are working on it, it might
be worth taking a look at what I had. I hadn't gotten to the part where I
needed to handle the progressive results on the callee side. I'll push what
I had in an hour or two and let you know which branch it is on.

**Comment by DZabavchik** [comment]
@davidchappelle, I just noticed that progressive and pattern-based registrations issue were unresolved for quite some time.  Was under impression that you are pushing the transport part forward (was hoping to see an example with websockets from you :)  No need for competition and duplicate work.  Please check pattern_registration_subs branch. It has a clean version of this failed PR with no windows stuff.  Let me know (I removed progressive calls provision from there)

**Comment by jpetso** [comment]
Another thing I wanted to mention, for the `intermediate` bool, is that it should likely be an enum instead of boolean. It looks nicer if I call, say,
```C++
invocation.result(std::tuple(), autobahn::result_progress);
```
than the one from the PR, which looks like this:
```C++
invocation.result(std::tuple(), true); // do I really have to add comments to write readable code?
```

**Comment by DZabavchik** [comment]
Missed the comment, sry. 

In that case, wouldn't explicit intermediate_result() or progress() that use private method internally be a better choice?

Regards,

Denis

**Comment by jpetso** [comment]
Actually yes, that would be much better API :)
Go for it!

**Comment by jpetso** [comment]
Personally, I like progress() better, not only is it shorter and more intuitive, but the name also matches up with what the specification calls it.

**Comment by DZabavchik** [comment]
@jpetso done
